### ISSUE TITLE
Envoy-stats target endpoint down in prometheus with istio 1.0.3

### ISSUE THREAD

**Issue by StefanCenusa** [istio/istio#9660]
**Describe the bug**
Since upgrading to istio 1.0.3, in prometheus envoy-stats target, some endpoints are *DOWN* with Error `err="no token found"`. I have found out that the reason of this error is invalid metrics being exported by that endpoints.

I am running an application which connects to a mongodb instance. I've got the metrics via istio-proxy (`<pod_ip>:15090/stats/prometheus`) as prometheus does. Then, with promtools I've check their validity. It failed because of these lines:
```
# TYPE envoy_mongo_outbound|27017||<my-mongodb>_collection_<my-collection>_query_total counter
envoy_mongo_outbound|27017||<my-mongodb>_collection_<my-collection>_query_total{} 20
```
I have a ServiceEntry for that mongodb which runs outside of the mesh.

My guess is that since **istio-statsd-prom-bridge** was removed, some metrics exposed by istio-proxy are not passing [Prometheus metrics validation](https://prometheus.io/docs/concepts/data_model/) (`[a-zA-Z_:][a-zA-Z0-9_:]*`) because they contain `|` chars. I've also tested this usecase in version 1.0.2 and all `|` are replaced with `_` by the statsd-prom-bridge.

**Expected behavior**
All endpoints to be UP for envoy-stats target.
Querying Prometheus to return `envoy_mongo_outbound|27017||<my-mongodb>_collection_<my-collection>_query_total` metrics.

**Steps to reproduce the bug**
Install istio 1.0.3.
Create a ServiceEntry for a mongodb instances (with protocol MONGO).
Create a service that connects to that mongo.
Open prometheus interface

**Version**
istio 1.0.3
kubernetes 1.11.3

**Installation**
Using helm chart
hub: docker.io/istio
tag: 1.0.3

**Comment by douglas-reid** [comment]
@StefanCenusa Thanks for reporting. I actually ran into this issue with other metrics when we made the switch, but unfortunately didn't have a mongo entry as part of the test cases.  This is the reason why we provide our own [stats_tags](https://github.com/istio/istio/blob/master/tools/deb/envoy_bootstrap_v2.json#L14-L43) for envoy in the bootstrap, as well as one reason why we [drop cluster-related metrics](https://github.com/istio/istio/blob/master/install/kubernetes/helm/subcharts/prometheus/templates/configmap.yaml#L61-L75) for the time being.

Would you be willing to try the following (ordered by difficulty, least to most):
1. adding a `drop` rule to the `metric_relabel_configs` for anything named `mongo_.*`
2. updating the `stats_tag` section with the appropriate regex for the mongo metric to create a label of `cluster` for it. This will require building, tagging, and pushing a new Docker image and then updating your injector configmap to use that image and deleting the running pod.

I can do both steps, but I don't have any way to verify their impact.

**Comment by StefanCenusa** [comment]
@douglas-reid Thanks for your reply.

For the first one, I'll add this rule to `metric_relabel_configs` and it should work.
```
- source_labels: [ __name__ ]
  regex: 'envoy_mongo.*'
  action: drop
```
I find the second step trickier because there multiple schemas possible for the [mongo proxy statistics](https://www.envoyproxy.io/docs/envoy/latest/configuration/network_filters/mongo_proxy_filter#statistics). I can't figure out a regex to match only the cluster part in every situation.
Here's an example of some possible metric formats:
```
mongo.outbound|27017||my-database.mongo.example.collection.app.query.total
mongo.outbound|27017||my-database.mongo.example.cmd.ismaster.reply_size
mongo.outbound|27017||my-database.mongo.example.cx_drain_close
```
where `my-database.mongo.example` is the database host.


Are you planning to restore the dropped metrics in future releases?

**Comment by douglas-reid** [comment]
@StefanCenusa We do not have any formal plans to add the dropped metrics at the moment. There are a few concerns (mainly around two things: scale and confusion(overlap with the istio_* metrics for instance) ) that need to be addressed, so at the moment it is basically on an as-needed basis. However, we are always open to requests, etc. 

Please let me know how it goes with the new drop rule, and we can follow up with a PR for the short-term fix. I can also help with the regex development (we can use multiple if needed).

**Comment by douglas-reid** [comment]
@StefanCenusa does #9974 provide you with an adequate regex example to unblock?

**Comment by StefanCenusa** [comment]
@douglas-reid Sorry for the delay, I had to downgrade to 1.0.2 in order to fix 503 issues caused by pilot, thus changing the setup mentioned above.

**Installation**:

```
hub: gcr.io/istio-release
tag: release-1.0-latest-daily
```

I've just tested that regex against istio-proxy reported stats. Although the envoy proxy docs says it would be `delay_injected`, the actual metric reported is `delays_injected`. 

Here's the full list:

```
mongo.outbound|27017||my-database.mongo.example.cmd.ismaster.total
mongo.outbound|27017||my-database.mongo.example.cmd.insert.total
mongo.outbound|27017||my-database.mongo.example.collection.proc.query.scatter_get
mongo.outbound|27017||my-database.mongo.example.collection.proc.query.total
mongo.outbound|27017||my-database.mongo.example.cx_destroy_local_with_active_rq
mongo.outbound|27017||my-database.mongo.example.cx_destroy_remote_with_active_rq
mongo.outbound|27017||my-database.mongo.example.cx_drain_close
mongo.outbound|27017||my-database.mongo.example.decoding_error
mongo.outbound|27017||my-database.mongo.example.delays_injected
mongo.outbound|27017||my-database.mongo.example.op_command
mongo.outbound|27017||my-database.mongo.example.op_command_reply
mongo.outbound|27017||my-database.mongo.example.op_get_more
mongo.outbound|27017||my-database.mongo.example.op_insert
mongo.outbound|27017||my-database.mongo.example.op_kill_cursors
mongo.outbound|27017||my-database.mongo.example.op_query
mongo.outbound|27017||my-database.mongo.example.op_query_active
mongo.outbound|27017||my-database.mongo.example.op_query_await_data
mongo.outbound|27017||my-database.mongo.example.op_query_exhaust
mongo.outbound|27017||my-database.mongo.example.op_query_multi_get
mongo.outbound|27017||my-database.mongo.example.op_query_no_cursor_timeout
mongo.outbound|27017||my-database.mongo.example.op_query_no_max_time
mongo.outbound|27017||my-database.mongo.example.op_query_scatter_get
mongo.outbound|27017||my-database.mongo.example.op_query_tailable_cursor
mongo.outbound|27017||my-database.mongo.example.op_reply
mongo.outbound|27017||my-database.mongo.example.op_reply_cursor_not_found
mongo.outbound|27017||my-database.mongo.example.op_reply_query_failure
mongo.outbound|27017||my-database.mongo.example.op_reply_valid_cursor
```

Except that, the regex matches everything else.

Moreover, I've added this rule to `metric_relabel_configs`, but the target is still `DOWN` with  `no token found`.

```
- source_labels: [ __name__ ]
  regex: 'envoy_mongo.*'
  action: drop
```

**Comment by douglas-reid** [comment]
@StefanCenusa with the merge of #9974, I believe this issue is now closed (in 1.1 and forward). Please re-open if you believe this is not the case.
